### PR TITLE
Signal: allow to use integral Enum for reset value.

### PR DESCRIPTION
**Pull request by Fatsie**
Follow-up to #294.

### CODE REVIEW

File: nmigen/hdl/ast.py [hunk 1]
```diff
@@ -834,6 +834,11 @@ def __init__(self, shape=None, *, name=None, reset=0, reset_less=False, min=None
                  attrs=None, decoder=None, src_loc_at=0):
         super().__init__(src_loc_at=src_loc_at)
 
+        if isinstance(reset, Enum):
+            reset = reset.value
+        if not type(reset) == int:
```
Comment (whitequark) [inline nmigen/hdl/ast.py]: This should be an `isinstance(reset, int)`, or `reset=True` will fail. (`bool` is derived from `int` in Python.)

File: nmigen/hdl/ast.py [hunk 2]
```diff
@@ -797,7 +797,7 @@ class Signal(Value, DUID):
         name this ``Signal`` is assigned to. Name collisions are automatically resolved by
         prepending names of objects that contain this ``Signal`` and by appending integer
         sequences.
-    reset : int
+    reset : int, bool or integral Enum
```
Comment (whitequark) [inline nmigen/hdl/ast.py]: Nit: `bool` is an integral type, so it's fine to just say `int or integral Enum`; this covers the cases where it's bool or bool Enum.

**Comment by whitequark** [comment]
Thanks. Could you add a test as well?

File: nmigen/hdl/ast.py (hunk 2 above)
Comment (Fatsie) [reply nmigen/hdl/ast.py]: I prefer to mention ``bool`` explicitly here as it makes clear that it is a proper use of the parameter.

**Comment by Fatsie** [comment]
Pushed a test.

File: nmigen/hdl/ast.py (hunk 2 above)
Comment (whitequark) [reply nmigen/hdl/ast.py]: This is inconsistent with all other docstrings nMigen has, which assume that the reader knows `bool` is an integral type.

Comment (Fatsie) [reply nmigen/hdl/ast.py]: I'll change it but it is not how I read autodocs. To me it's not about knowing if ``bool`` is an integral type but if ``int`` as a specification in docstring means one is also supposed to pass  ``bool`` for that parameter. To me just specifying ``int`` means one is not supposed use ``bool`` for that parameter.

**Comment by Fatsie** [comment]
Pushed change that removes reference to bool in docstring and extended test to check for non-integral Enum type.

File: nmigen/hdl/ast.py (hunk 2 above)
Comment (whitequark) [reply nmigen/hdl/ast.py]: That's not how object-oriented programming works. If `bool` derives from `int`, then it is OK to pass `bool` anywhere you'd pass `int`; this is called the Liskov Substitution Principle. Over-explaining LSP in a docstring is like over-explaining classes in a docstring: sure, you could do it, but why?

File: nmigen/hdl/ast.py [hunk 3]
```diff
@@ -834,6 +834,11 @@ def __init__(self, shape=None, *, name=None, reset=0, reset_less=False, min=None
                  attrs=None, decoder=None, src_loc_at=0):
         super().__init__(src_loc_at=src_loc_at)
 
+        if isinstance(reset, Enum):
+            reset = reset.value
+        if not isinstance(reset, int):
+            raise ValueError("Reset has to be an int or an integral Enum")
```
Comment (whitequark) [inline nmigen/hdl/ast.py]: This should be a `TypeError`, and talk about reset *value*, not just reset, which is a different concept.

**Comment by Fatsie** [comment]
``ValueError`` -> ``TypeError``

File: nmigen/hdl/ast.py (hunk 3 above)
Comment (Fatsie) [reply nmigen/hdl/ast.py]: Done.

**Comment by whitequark** [comment]
Thanks for the PR!

**Pull request merged**